File: custom_diamond_app/custom_diamond_app/report/bank_account/bank_account.py
# Copyright (c) 2022, kiran and contributors
# For license information, please see license.txt

import frappe
from frappe import utils
from frappe.utils import today
from frappe import _, scrub
from frappe.utils import cint, flt
from six import iteritems

from erpnext.accounts.party import get_partywise_advanced_payment_amount
from erpnext.accounts.report.accounts_receivable.accounts_receivable import ReceivablePayableReport
from erpnext.accounts.doctype.bank_account import bank_account
from datetime import date

def execute(filters=None):
	args = {
		"party_type": "Customer",
		"naming_by": ["Selling Settings", "cust_master_name"],
	}

	return AccountsReceivableSummary(filters).run(args)


class AccountsReceivableSummary(ReceivablePayableReport):
	def run(self, args):
		self.party_type = args.get("party_type")
		self.party_naming_by = frappe.db.get_value(
			args.get("naming_by")[0], None, args.get("naming_by")[1]
		)
		
		self.get_columns()
		self.get_data(args)
		# print(self.data,"/////////////////////////////////////////")
		return self.columns, self.data

	def get_data(self, args):
		self.data = []
		self.receivables = ReceivablePayableReport(self.filters).run(args)[1]
		self.get_party_total(args)
		party_advance_amount = (
			get_partywise_advanced_payment_amount(
				self.party_type,
				self.filters.report_date,
				self.filters.show_future_payments,
				self.filters.company,
			)
			or {}
		)
		

		if self.filters.show_gl_balance:
			gl_balance_map = get_gl_balance(self.filters.report_date)

		for party, party_dict in iteritems(self.party_total):
			if party_dict.outstanding == 0:
				continue

			row = frappe._dict()
			
			row.party = party
			if self.party_naming_by == "Naming Series":
				row.party_name = frappe.get_cached_value(
					self.party_type, party, scrub(self.party_type) + "_name"
				)
			acc = frappe.db.sql(
				"""select account_name,bank,branch_code,bank_account_no as bank_account_number,party from `tabBank Account` where party = '{0}'
				""".format(party),
				as_dict=1,	
				)
			if acc:
				row.update(acc[0])

			row.update(party_dict)

			# Advance against party
			row.advance = party_advance_amount.get(party, 0)
			# export_date = []
			if self.filters.show_gl_balance:
				row.gl_balance = gl_balance_map.get(party)
				row.diff = flt(row.outstanding) - flt(row.gl_balance)
				# row.export_date = frappe.utils.formatdate(today(), "dd-MM-yyyy")
				# print(export_date,"000000000000000000000000")
			self.data.append(row)

	def get_party_total(self, args):
		self.party_total = frappe._dict()
		for d in self.receivables:
			self.init_party_total(d)


			# Add all amount columns
			for k in list(self.party_total[d.party]):
				if k not in ["currency", "sales_person"]:
					self.party_total[d.party][k] += d.get(k, 0.0)


	def init_party_total(self, row):
		self.party_total.setdefault(
			row.party,
			frappe._dict(
				{
					"outstanding": 0.0,
				},
			),
		)
                                        	
	def get_columns(self):
		self.columns = []
		if self.party_naming_by == "Naming Series":
			self.add_column(_("Bank Account Number"),fieldname="bank_account_number", fieldtype="Int")
			self.add_column(_("Amount"), fieldname="outstanding")
			self.add_column(_("Account Name"),fieldname="account_name", fieldtype="Link")
			self.add_column(_("{0} Name").format(self.party_type), fieldname="party_name", fieldtype="Data")
			self.add_column(label=_(self.party_type),fieldname="party",fieldtype="Link",
						options=self.party_type,width=180,)
			self.add_column(_("Export date"))
			self.add_column(_("Branch Code"), fieldname="branch_code", fieldtype="Data")
			self.add_column(_("Bank"),fieldname="bank", fieldtype="Data")
			# self.add_column(_("Email Id"), fieldname="email_id", fieldtype="Data")
			# self.add_column(_("Account Name"),fieldname="account_name", fieldtype="Data")

		if self.filters.show_sales_person:
			self.add_column(label=_("Sales Person"), fieldname="sales_person", fieldtype="Data")
		

def get_gl_balance(report_date):
	return frappe._dict(
		frappe.db.get_all(
			"GL Entry",
			fields=["party", "sum(debit -  credit)"],
			filters={"posting_date": ("<=", report_date), "is_cancelled": 0},
			group_by="party",
			as_list=1,
		)
	)

def execute(filters=None):
	args = {
		"party_type": "Supplier",
		"naming_by": ["Buying Settings", "supp_master_name"],
	}
	return AccountsReceivableSummary(filters).run(args)

File: custom_diamond_app/custom_diamond_app/report/bank_account/bank_account.js
// Copyright (c) 2022, Ganu Reddy and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Bank Account"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company")
		},
		{
			"fieldname":"report_date",
			"label": __("Posting Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"ageing_based_on",
			"label": __("Ageing Based On"),
			"fieldtype": "Select",
			"options": 'Posting Date\nDue Date',
			"default": "Due Date"
		},
		{
			"fieldname":"range1",
			"label": __("Ageing Range 1"),
			"fieldtype": "Int",
			"default": "30",
			"reqd": 1
		},
		{
			"fieldname":"range2",
			"label": __("Ageing Range 2"),
			"fieldtype": "Int",
			"default": "60",
			"reqd": 1
		},
		{
			"fieldname":"range3",
			"label": __("Ageing Range 3"),
			"fieldtype": "Int",
			"default": "90",
			"reqd": 1
		},
		{
			"fieldname":"range4",
			"label": __("Ageing Range 4"),
			"fieldtype": "Int",
			"default": "120",
			"reqd": 1
		},
		{
			"fieldname":"finance_book",
			"label": __("Finance Book"),
			"fieldtype": "Link",
			"options": "Finance Book"
		},
		{
			"fieldname":"cost_center",
			"label": __("Cost Center"),
			"fieldtype": "Link",
			"options": "Cost Center",
			get_query: () => {
				var company = frappe.query_report.get_filter_value('company');
				return {
					filters: {
						'company': company
					}
				}
			}
		},
		{
			"fieldname":"supplier",
			"label": __("Supplier"),
			"fieldtype": "Link",
			"options": "Supplier"
		},
		{
			"fieldname":"payment_terms_template",
			"label": __("Payment Terms Template"),
			"fieldtype": "Link",
			"options": "Payment Terms Template"
		},
		{
			"fieldname":"supplier_group",
			"label": __("Supplier Group"),
			"fieldtype": "Link",
			"options": "Supplier Group"
		},
		{
			"fieldname":"based_on_payment_terms",
			"label": __("Based On Payment Terms"),
			"fieldtype": "Check",
		},
		{
			"fieldname":"empty_columns",
			"label": __("Empty Columns"),
			"fieldtype": "Check",
			"default": 0,
		},
		{
			"fieldname":"no_of_empty_columns",
			"label": __("No of Empty Columns"),
			"fieldtype":"Data",
			// "hidden":1

		}
	],

	onload: function(report) {
		report.page.add_inner_button(__("Accounts Payable"), function() {
			var filters = report.get_values();
			frappe.set_route('query-report', 'Accounts Payable', {company: filters.company});
		});


		document.querySelectorAll("[data-fieldname='no_of_empty_columns'] input")[0].style.display='none' 
		let empty_column = document.querySelectorAll("[data-fieldname='empty_columns'] input")[0]
		empty_column.addEventListener('change',(event)=>{
			let empty_columns = frappe.query_report.get_filter_value('empty_columns');
			if(empty_columns){
				document.querySelectorAll("[data-fieldname='no_of_empty_columns'] input")[0].style.display='block' 
			}else{
				document.querySelectorAll("[data-fieldname='no_of_empty_columns'] input")[0].style.display='none' 
			}
		})

	}
};

erpnext.utils.add_dimensions('Bank Account', 9);



